Implement optimal member selection

AnnikaLau opened the pull request.

huppd reviewed (commented):
Thanks a lot for the neat new feature. I added some remarks for the code. I think it would be also great to have some u-tests for the new functionality.

AnnikaLau commented on engine/select_optimal_members.py:
```diff
@@ -0,0 +1,305 @@
+            logger.info("Try with {} members.".format(mem_num))
+            max_passed = 1
+            vars = []
+            for i in range(1, 51):
```
With your above suggestions, `iter` starts at `0`. However, I find it nicer when it is the actual number of the iteration. Right now it gives the output "Test 1,... Test 2...." etc. Also the calculations for stopping earlier is based on `iter` starting at `1`. Of course, I could adapt it there instead but this would be less nice I think.

huppd replied:
I think using starting index 1 in a zero based language makes the code more complicated as needed. And for iterations I am really used to have them starting at zero. For the output you can just add a +1 there and adjust the criterion.

AnnikaLau commented on engine/select_optimal_members.py:
```diff
@@ -0,0 +1,305 @@
+    max_factor,
+):
+
+    members = [i for i in range(1, total_member_num + 1)]
```
Hmm, I think it's nicer to have the members starting at 1. Also this would mean a lot of changes, including updating all optimal member selections in ICON in [set_probtest_ensemble_num](run/tolerance/set_probtest_ensemble_num), as they are based on the members starting at 1...

huppd replied:
Again, I think using starting index 1 in a zero based language makes the code more complicated as needed. But for the members I leave it up to you.

AnnikaLau replied:
Then I decide for the members to leave it like this ;-)

AnnikaLau commented on engine/select_optimal_members.py:
```diff
@@ -0,0 +1,305 @@
+            stats_file_name.format(member_id=m_id), index_col=[0, 1, 2]
+        )
+
+        # check if variables are available in reference file
```
I added the function `tolerance_test` to the `check.py` script: https://github.com/MeteoSwiss/probtest/pull/26/commits/7f8774ec3e4eeb32f56fe261a14bd2284b6ebbf8

What do you think?

huppd replied:
I think it is great. I only have two minor suggestions :upside_down_face: I would locate the function if possible somewhere in utils (maybe dataframe_ops.py or utils.py?). The I would rename it. I like when functions are named by verbs or start with a verb, so I would go with something like `test_stats_files_with_tolerances` or `test_with_tolerances`.

AnnikaLau replied:
The current function `tolerance_test` is using the function `check_intersection` and `check_variable`, all in the `check.py` script. Would you recommend to move all three functions then?

huppd replied:
Yes, I think it would make sense to move them as well and I think dataframe_ops is probably then the right location. I think this would make the code structure a bit more cleaner.

AnnikaLau replied:
I moved all three functions and renamed `tolerance_test` to `test_stats_file_with_tolerances`

huppd replied:
Thanks.

AnnikaLau commented:
> Thanks a lot for the neat new feature. I added some remarks for the code. I think it would be also great to have some u-tests for the new functionality. A bit nitpicking but it not taking the optimal selection but a good robust one, so the name might be misleading.

@huppd Thanks a lot for the review!
You are right `select_optimal_members` is maybe not the optimal name ;-). What do you think of `select_stable_members` or `select_robust_members` instead?

Also I added some unittests now: https://github.com/MeteoSwiss/probtest/blob/optimal_member_selection/tests/engine/test_select_optimal_members.py

huppd commented:
> > Thanks a lot for the neat new feature. I added some remarks for the code. I think it would be also great to have some u-tests for the new functionality. A bit nitpicking but it not taking the optimal selection but a good robust one, so the name might be misleading.
> 
> @huppd Thanks a lot for the review! You are right `select_optimal_members` is maybe not the optimal name ;-). What do you think of `select_stable_members` or `select_robust_members` instead?
> 
> Also I added some unittests now: https://github.com/MeteoSwiss/probtest/blob/optimal_member_selection/tests/engine/test_select_optimal_members.py

Thanks for the tests. They look good to me. It is great to have them and also in a way that is not directly relying on ICON runs :)
My favorite name would be probably `create_member_selection_with_adequate_tolerances`. Maybe `create_member_selection` might be good enough?

AnnikaLau commented on util/click_util.py:
```diff
@@ -97,6 +97,16 @@ def convert(self, value, param, ctx):
+    "test_tolerance": r"Test with how many stats files out "
+    + r"of x (x=total_member_num) the probtest passes given a specific tolerance file.",
+    "total_member_num": r"Number of ensemble members used "
+    + r"to select the x optimal members from (min_member_num<=x<=max_member_num).",
```
Hmm as a user I would find it more intuitive that the `max_member_num` is the actual maximum number..

huppd replied:
Also fine for me.

AnnikaLau commented:
> > > Thanks a lot for the neat new feature. I added some remarks for the code. I think it would be also great to have some u-tests for the new functionality. A bit nitpicking but it not taking the optimal selection but a good robust one, so the name might be misleading.
> > 
> > 
> > @huppd Thanks a lot for the review! You are right `select_optimal_members` is maybe not the optimal name ;-). What do you think of `select_stable_members` or `select_robust_members` instead?
> > Also I added some unittests now: https://github.com/MeteoSwiss/probtest/blob/optimal_member_selection/tests/engine/test_select_optimal_members.py
> 
> Thanks for the tests. They look good to me. It is great to have them and also in a way that is not directly relying on ICON runs :) My favorite name would be probably `create_member_selection_with_adequate_tolerances`. Maybe `create_member_selection` might be good enough?

Considering that this name is also used as a command, i.e. `python probtest.py select-optimal-members`, the name should probably rather by shorter than longer. The other function names used as a command are rather short (`init`, `stats`, `tolerance`, etc.). Wouldn't it be better to rename it to `select-members` and add good documentation strings instead?

huppd commented:
> > > > Thanks a lot for the neat new feature. I added some remarks for the code. I think it would be also great to have some u-tests for the new functionality. A bit nitpicking but it not taking the optimal selection but a good robust one, so the name might be misleading.
> > > 
> > > 
> > > @huppd Thanks a lot for the review! You are right `select_optimal_members` is maybe not the optimal name ;-). What do you think of `select_stable_members` or `select_robust_members` instead?
> > > Also I added some unittests now: https://github.com/MeteoSwiss/probtest/blob/optimal_member_selection/tests/engine/test_select_optimal_members.py
> > 
> > 
> > Thanks for the tests. They look good to me. It is great to have them and also in a way that is not directly relying on ICON runs :) My favorite name would be probably `create_member_selection_with_adequate_tolerances`. Maybe `create_member_selection` might be good enough?
> 
> Considering that this name is also used as a command, i.e. `python probtest.py select-optimal-members`, the name should probably rather by shorter than longer. The other function names used as a command are rather short (`init`, `stats`, `tolerance`, etc.). Wouldn't it be better to rename it to `select-members` and add good documentation strings instead?

Yes, you are right. I think `select-members` is the better name.

AnnikaLau commented:
@huppd I implemented all your requested changes. Do you want to have a look at it again?
